app/app.go

```diff
@@ -521,26 +593,37 @@ func (app *App) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) abci.R
 // EndBlocker application updates every end block
 func (app *App) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
 	res := app.mm.EndBlock(ctx, req)
-	// NOTE: this is a specific feature for upgrading from v1 to v2. It will be deprecated in v3
-	if app.UpgradeKeeper.ShouldUpgradeToV2(req.Height) && app.AppVersion(ctx) == v1.Version {
-		if err := app.Upgrade(ctx, v2.Version); err != nil {
-			panic(err)
+	currentVersion := app.AppVersion()
+	// For v1 only we upgrade using a agreed upon height known ahead of time
+	if currentVersion == v1 {
+		// check that we are at the height before the upgrade
+		if req.Height == app.upgradeHeight-1 {
+			if err := app.Upgrade(ctx, currentVersion, currentVersion+1); err != nil {
+				panic(err)
+			}
 		}
 		// from v2 to v3 and onwards we use a signalling mechanism
-	} else if shouldUpgrade, version := app.UpgradeKeeper.ShouldUpgrade(); shouldUpgrade {
+	} else if shouldUpgrade, newVersion := app.UpgradeKeeper.ShouldUpgrade(); shouldUpgrade {
 		// Version changes must be increasing. Downgrades are not permitted
-		if version > app.AppVersion(ctx) {
-			if err := app.Upgrade(ctx, version); err != nil {
+		if newVersion > currentVersion {
+			if err := app.Upgrade(ctx, currentVersion, newVersion); err != nil {
 				panic(err)
 			}
 		}
 	}
 	return res
 }
 
-func (app *App) Upgrade(ctx sdk.Context, version uint64) error {
-	app.SetAppVersion(ctx, version)
-	return app.mm.RunMigrations(ctx, app.configurator, app.AppVersion(ctx), version)
+func (app *App) Upgrade(ctx sdk.Context, fromVersion, toVersion uint64) error {
+	if err := app.mm.RunMigrations(ctx, app.configurator, fromVersion, toVersion); err != nil {
+		return err
+	}
+	if toVersion == v2 {
+		// we need to set the app version in the param store for the first time
```

Comment:
The crux was that since we pass the sdk context everywhere, we should use that to get the app version. The app version should technically be present in the context because it includes the header that has the app version. The problem however was that in most cases when we create the context we only populate parts of the header (the height for example). So I needed to make several changes in `PrepareProposal`, `ProcessProposal`, `InitGenesis` etc. that we actually populated the app version